File: companies/Societe Generale/17.letter-combinations-of-a-phone-number.java
import java.util.*;

/**
 * Algorithm:
 * Backtrack through the Cartesian product of letters for each digit. The path
 * has one chosen character per processed digit; when it reaches full length,
 * it becomes one answer.
 *
 * Java data structures:
 * A String[] maps digit characters to letter strings. StringBuilder is the
 * mutable path.
 *
 * Complexity:
 * Time O(4^n * n), recursion space O(n), plus output.
 */
class Solution {
    private static final String[] PHONE = {
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
    };
    private List<String> ans;
    private String digits;
    private StringBuilder path;

    public List<String> letterCombinations(String digits) {
        ans = new ArrayList<>();
        if (digits == null || digits.isEmpty()) {
            return ans;
        }
        this.digits = digits;
        path = new StringBuilder();
        backtrack(0);
        return ans;
    }

    private void backtrack(int index) {
        if (index == digits.length()) {
            ans.add(path.toString());
            return;
        }
        String letters = PHONE[digits.charAt(index) - '0'];
        for (int i = 0; i < letters.length(); i++) {
            path.append(letters.charAt(i));
            backtrack(index + 1);
            path.deleteCharAt(path.length() - 1);
        }
    }
}

File: companies/Tesla/17.letter-combinations-of-a-phone-number.go
package leetcode

// LetterCombinations17 backtracks through the Cartesian product of letters for
// each digit. A byte path is reused during recursion and copied into a string at
// each leaf.
//
// Time: O(4^n*n). Space: O(n) recursion excluding output.
func LetterCombinations17(digits string) []string {
	if digits == "" {
		return []string{}
	}
	phone := map[byte]string{'2': "abc", '3': "def", '4': "ghi", '5': "jkl", '6': "mno", '7': "pqrs", '8': "tuv", '9': "wxyz"}
	ans := []string{}
	path := []byte{}
	var dfs func(int)
	dfs = func(idx int) {
		if idx == len(digits) {
			ans = append(ans, string(path))
			return
		}
		for i := 0; i < len(phone[digits[idx]]); i++ {
			path = append(path, phone[digits[idx]][i])
			dfs(idx + 1)
			path = path[:len(path)-1]
		}
	}
	dfs(0)
	return ans
}
